src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/HeapOfActiveAtoms.java

```diff
@@ -0,0 +1,270 @@
+	 */
+	public void newNoGoods(Collection<NoGood> newNoGoods) {
+		for (NoGood newNoGood : newNoGoods) {
+			Type type = newNoGood.getType();
+			if (type != Type.LEARNT && type != Type.INTERNAL) {
+				analyzeNewNoGood(newNoGood);
+				initActivity(newNoGood);
+			}
+		}
+	}
+	
+	/**
+	 * May be implemented in subclasses to add specific analysis of nogoods.
+	 */
+	protected void analyzeNewNoGood(NoGood newNoGood) {
+	}
+
+	/**
+	 * Computes and stores initial activity values for the atoms occurring in the given nogood.
+	 */
+	protected void initActivity(NoGood newNoGood) {
+		if (moms != null) {
+			initActivityMOMs(newNoGood);
+		} else {
+			initActivityNaive(newNoGood);
+		}
+	}
+
+	/**
+	 * Uses {@link MOMs} to initialize activity scores, which are then scaled to the interval [0,1].
+	 * This is done by computing 1 - 1/log(s+1.01) for original score s.
+	 * This guarantees a normalized score between 0 and 1 and retains relative order.
+	 * 1.01 is added to avoid computing the logarithm of a number between 0 and 1 (input scores have to be greater or equal to 0!)
+	 * @param newNoGood a new nogood, the atoms occurring in which will be initialized
+	 */
+	private void initActivityMOMs(NoGood newNoGood) {
+		LOGGER.debug("Initializing activity scores with MOMs");
+		for (int literal : newNoGood) {
+			int atom = atomOf(literal);
+			if (atom >= initializedActivityScores.length || !initializedActivityScores[atom]) {
+				double score = moms.getScore(atom);
+				if (score > 0.0) {
+					incrementActivity(atom, 1 - 1 / (Math.log(score + 1.01)));
+				}
+			}
+		}
+	}
+
+	private void growForMaxAtomId(int maxAtomId) {
+		// Grow arrays only if needed.
+		if (activityScores.length > maxAtomId) {
+			return;
+		}
+		// Grow by default growth factor, except if bigger array is required due to maxAtomId.
+		int newCapacity = arrayGrowthSize(activityScores.length);
+		if (newCapacity < maxAtomId + 1) {
+			newCapacity = maxAtomId + 1;
+		}
+
+		activityScores = Arrays.copyOf(activityScores, newCapacity);
+		initializedActivityScores = Arrays.copyOf(initializedActivityScores, newCapacity);
+	}
+
+	private void initActivityNaive(NoGood newNoGood) {
+		LOGGER.debug("Initializing activity scores naively");
+		for (Integer literal : newNoGood) {
+			incrementActivity(atomOf(literal));
+		}
+	}
+
+	/**
+	 * Returns the atom with the highest activity score and removes it from the heap.
+	 */
+	public Integer getMostActiveAtom() {
+		return heap.poll();
+	}
+
+	/**
+	 * Increments the activity of the given atom
+	 * 
+	 * by adding to it the current activity increment times the increment factor.
+	 * If the new value exceeds a certain threshold, all activity scores are normalized.
+	 */
+	public void incrementActivity(int atom) {
+		incrementActivity(atom, currentActivityIncrement);
+	}
+	
+	protected void incrementActivity(int atom, double increment) {
+		growForMaxAtomId(atom);
```

Comment:
I think this `growForMaxAtomId` needs a more principled solution, similarly as it was necessary for the `Assignment` some time ago, where at first I also tried to detect the max atom id locally (and which also was buggy in the beginning).
Now the `AtomStore` provides a `getMaxAtomId()` method that returns a correct value and the `DefaultSolver.ingest(obtainedNoGoods)` method calls the `growForMaxAtomId()` on the assignment and the `NoGoodStore`. I think there should now be an additional call for the heuristics to grow. So this growing is done at one place for all components of the solver.